### ISSUE TITLE
20% testing accuracy difference CPU and GPU in CIFAR10 dataset

### ISSUE THREAD

**Issue by peerajak** [BVLC/caffe#2624]
CPU and GPU traning should have given the same result, but no!
-------------------------------------------
I am training Cifar10 example with my own architecture shown below. I think it is suppose to be that CPU or GPU would give roughly the same accuracy and loss values.
With all other parameters being the same except CPU or GPU, my two experiment has much different test accuracy. 

GPU: 
I0619 12:18:22.621153 15991 solver.cpp:269] Iteration 5000, Testing net (#0)
I0619 12:18:22.898356 15991 solver.cpp:318]     Test net output #0: accuracy = 0.5712
I0619 12:18:22.898421 15991 solver.cpp:318]     Test net output #1: loss = 1.35011 (* 1 = 1.35011 loss)

CPU:
I0619 12:28:27.662317 15616 solver.cpp:269] Iteration 5000, Testing net (#0)
I0619 12:28:31.197854 15616 solver.cpp:318]     Test net output #0: accuracy = 0.3764
I0619 12:28:31.197901 15616 solver.cpp:318]     Test net output #1: loss = 1.78204 (* 1 = 1.78204 loss)

Here is my solver.prototxt
---------------------------------------------
net: "examples/cifar10/cifar10_cs231n_train_test.prototxt"
test_iter: 100
test_interval: 500
base_lr: 0.0001
momentum: 0.9
weight_decay: 0.001
lr_policy: "fixed"
display: 100
max_iter: 5000
snapshot: 500
snapshot_prefix: "examples/cifar10/cifar10_cs231n"
solver_mode: GPU

-----------------------------------------------

Here is my train_test.prototxt
-------------------------------------------------------------
name: "CIFAR10_cs231n"
layers {
  name: "cifar"
  type: DATA
  top: "data"
  top: "label"
  data_param {
    source: "examples/cifar10/cifar10_train_lmdb"
    batch_size: 100
    backend: LMDB
  }
  transform_param {
    mean_file: "examples/cifar10/mean.binaryproto"
  }
  include: { phase: TRAIN }
}
layers {
  name: "cifar"
  type: DATA
  top: "data"
  top: "label"
  data_param {
    source: "examples/cifar10/cifar10_test_lmdb"
    batch_size: 100
    backend: LMDB
  }
  transform_param {
    mean_file: "examples/cifar10/mean.binaryproto"
  }
  include: { phase: TEST }
}
layers {
  name: "conv1"
  type: CONVOLUTION
  bottom: "data"
  top: "conv1"
  blobs_lr: 1
  blobs_lr: 2
  convolution_param {
    num_output: 32
    pad: 2
    kernel_size: 5
    stride: 1
    weight_filler {
      type: "constant"
      value: 0.001
    }
    bias_filler {
      type: "constant"
      value: 0.001
    }
  }
}
layers {
  name: "relu1"
  type: RELU
  bottom: "conv1"
  top: "relu1"
}
layers {
  name: "pool1"
  type: POOLING
  bottom: "relu1"
  top: "pool1"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layers {
  name: "ip1"
  type: INNER_PRODUCT
  bottom: "pool1"
  top: "ip1"
  blobs_lr: 1
  blobs_lr: 2
  inner_product_param {
    num_output: 10
    weight_filler {
      type: "constant"
      value: 0.001
    }
    bias_filler {
      type: "constant"
      value: 0.001
    }
  }
}
layers {
  name: "accuracy"
  type: ACCURACY
  bottom: "ip1"
  bottom: "label"
  top: "accuracy"
  include: { phase: TEST }
}
layers {
  name: "loss"
  type: SOFTMAX_LOSS
  bottom: "ip1"
  bottom: "label"
  top: "loss"
}
-----------------------------------------------

**Comment by peerajak** [comment]
Answer: Setting weight to constant values may not be read by GPU version
----------------------------------------------

It seems that the problem is about setting weights and bias to constant value is not read by GPU version.

After setting the weight to Gaussian with std 0.001, things solve. GPU and CPU now looks like previous GPU version's accuracy. Here it is.

GPU:
I0619 14:12:49.797749  2624 solver.cpp:251] Iteration 5000, loss = 0.853291
I0619 14:12:49.797785  2624 solver.cpp:269] Iteration 5000, Testing net (#0)
I0619 14:12:50.061450  2624 solver.cpp:318]     Test net output #0: accuracy = 0.5832
I0619 14:12:50.061511  2624 solver.cpp:318]     Test net output #1: loss = 1.3164 (* 1 = 1.3164 loss)
CPU:
I0619 14:27:02.443125  8895 solver.cpp:251] Iteration 5000, loss = 0.855041
I0619 14:27:02.443400  8895 solver.cpp:269] Iteration 5000, Testing net (#0)
I0619 14:27:07.659747  8895 solver.cpp:318]     Test net output #0: accuracy = 0.5731
I0619 14:27:07.659881  8895 solver.cpp:318]     Test net output #1: loss = 1.3771 (* 1 = 1.3771 loss)

**Comment by jshfeng** [comment]
This tracker is basically used for reporting bugs and issues with caffe. To ask questions about implementations, please refer to the mail list: https://groups.google.com/forum/#!forum/caffe-users
